### PR TITLE
Failure detector doc implementation details

### PR DESCRIPTION
**Pull request by xnull**
## Overview

Description:
Failure detector implementation details. Documentation 

Why should this be merged: 

Related issue(s) (if applicable): #<number>

## Checklist (Definition of Done):

- [x] There are no TODOs left in the code
- [x] [Coding conventions](https://github.com/CorfuDB/CorfuDB/wiki/Corfu-Style-Guidelines) (e.g. for logging, unit tests) have been followed
- [x] Change is covered by automated tests
- [x] Public API has Javadoc

### CODE REVIEW

File: docs/failure-detector/failure-detector.md [hunk 1]
```diff
@@ -49,4 +49,76 @@ In the example:
...
+ Node A believes that node B is disconnected <br/>
+ Node B believes that node A is connected <br/>
+ The graph will be changed to: <br/>
+ {a: [{a: OK, b: FAILED}]} <br/>
+ {b: [{a: FAILED, b: OK}]} <br/>
+ Node B is not connected to node A anymore.
+ 
+**Decision maker** - is a node which can decide to add a failed node into the unresponsive list.
+The decision maker is an technique, used to reduce parallel updates of the cluster state.
+Still, each node can choose its own decision maker.
```
Comment (lixinchengdu) [inline docs/failure-detector/failure-detector.md]: Why "each node can choose its own decision marker"? Does it mean each node choose the decision maker based on its own cluster view such that the decision maker meet all the requirements list below?

File: docs/failure-detector/failure-detector.md [hunk 2]
```diff
@@ -49,4 +49,76 @@ In the example:
...
+Note: it's not required to choose a particular one for entire cluster to add a node to unresponsive list.
+
+The decision maker must have:
+- highest number of successful connections in the graph.
+There could be many nodes with same number of successful connections, 
+then the decision maker will be a node with smallest name.
+
+We also have additional checks to prevent all possible incorrect ways. Decision maker can't be found if:
+- `ClusterGraph` is empty, which is an invalid state.
+- the decision maker doesnt have connections, which is also impossible.
```
Comment (lixinchengdu) [inline docs/failure-detector/failure-detector.md]: the decision maker doesn't

File: docs/failure-detector/failure-detector.md [hunk 3]
```diff
@@ -49,4 +49,76 @@ In the example:
...
+
+Detecting **failed node** - find a node in the cluster with a minimum number of successful connections.
```
Comment (lixinchengdu) [inline docs/failure-detector/failure-detector.md]: Is it possible to detect multiple failed nodes simultaneously?

File: docs/failure-detector/failure-detector.md [hunk 4]
```diff
@@ -49,4 +49,76 @@ In the example:
...
+- Descending order of number of connections
+- Ascending order of a node name (alphabetically)
+
+A `Decision maker` always has highest number of successful connections and smallest name and
+a failed node always has lowest number of successful connections and highest name.
+
+Finding a failed node - collect all node ranks in a graph and choose the node 
+with the smallest number of successful connections and with the highest name (sorted alphabetically).
+
+Even if we have found a node with smallest number of connections it doesnt mean that the node is
```
Comment (lixinchengdu) [inline docs/failure-detector/failure-detector.md]: What is the meaning of fully connected here? Doesn't fully connected means connect to every other nodes?

File: docs/failure-detector/failure-detector.md (hunk 1 above)
Comment (lixinchengdu) [reply docs/failure-detector/failure-detector.md]: Also, what if different decision markers have different decisions on failed nodes?

Comment (xnull) [reply docs/failure-detector/failure-detector.md]: 1. Yes you are right each node choose the decision maker based on its own cluster view 
2. It's possible that two different nodes decide to be a decision maker at the same time and can have different decisions on failed nodes. In that case, nodes would be added into the unresponsive list. On the next iteration, nodes would have a more consistent view on the cluster state and a node accidentally added to the unresponsive list will be healed

File: docs/failure-detector/failure-detector.md (hunk 3 above)
Comment (xnull) [reply docs/failure-detector/failure-detector.md]: yes, but the probability pretty low

File: docs/failure-detector/failure-detector.md (hunk 4 above)
Comment (xnull) [reply docs/failure-detector/failure-detector.md]: means - the node has connected to all responsive nodes in the cluster

File: docs/failure-detector/failure-detector.md (hunk 1 above)
Comment (xnull) [reply docs/failure-detector/failure-detector.md]: 1. Yes you are right each node choose the decision maker based on its own cluster view

2. It's possible that two different nodes decide to be a decision maker at the same time and can have different decisions on failed nodes. In that case, nodes would be added into the unresponsive list. On the next iteration, nodes would have a more consistent view on the cluster state and a node accidentally added to the unresponsive list will be healed

File: docs/failure-detector/failure-detector.md (hunk 2 above)
Comment (xnull) [reply docs/failure-detector/failure-detector.md]: done

File: docs/failure-detector/failure-detector.md (hunk 3 above)
Comment (xnull) [reply docs/failure-detector/failure-detector.md]: Sorry, actually we detected a failed node at a time, more than one node can't be added to the unresponsive list at a time

File: docs/failure-detector/failure-detector.md [hunk 5]
```diff
@@ -49,4 +49,76 @@ In the example:
...
+The healing detector heals nodes which were previously marked unresponsive but have now healed.
+
+
+**Collect poll report**
+Each poll round consists of iterations. In each iteration, the FailureDetector pings all responsive nodes 
+in the layout and also collects their node states to provide cluster state.
+To provide more resilient failre detection mechanism, the failure detector collects few "poll reports" 
+equal to `failureThreshold` number then aggregates final poll report
+
+Once poll report collected, the `failureDetectionTask` is triggered by Remote Monitoring Service.
```
Comment (hisundar) [inline docs/failure-detector/failure-detector.md]: nit: Once a poll report is collected,

File: docs/failure-detector/failure-detector.md [hunk 6]
```diff
@@ -49,4 +49,76 @@ In the example:
...
+
+Finding a failed node - collect all node ranks in a graph and choose the node 
+with the smallest number of successful connections and with the highest name (sorted alphabetically).
+
+Even if we have found a node with smallest number of connections it doesnt mean that the node is 
+failed. We also have to check if the node is *fully connected*. 
+Only if the node is not fully connected and has the smallest number of successful connections then
+the node can be added to the unresponsive list.   
+
+**Healing process** - determining if a node is alive and remove it from the unresponsive list.
```
Comment (hisundar) [inline docs/failure-detector/failure-detector.md]: nit: determine if a node is alive and remove it from the unresponsive list.

File: docs/failure-detector/failure-detector.md [hunk 7]
```diff
@@ -49,4 +49,76 @@ In the example:
  - node `a` is a decision maker, it has 3 successful connections
  - node `c` is a failed node, it has same number of successful connections like node `b` but is has higher node name.
  - node `a` will add node `c` to the unresponsive list.     
- 
\ No newline at end of file
+ 
+### Implementation
+
+**Remote Monitoring Service** constitutes of failure and healing monitoring and handling.
```
[the remark this answers is not in the document]
Comment (xnull) [reply docs/failure-detector/failure-detector.md]: yes)
fixed

File: docs/failure-detector/failure-detector.md (hunk 5 above)
Comment (xnull) [reply docs/failure-detector/failure-detector.md]: fixed

File: docs/failure-detector/failure-detector.md (hunk 6 above)
Comment (xnull) [reply docs/failure-detector/failure-detector.md]: fixed

File: docs/failure-detector/failure-detector.md [hunk 8]
```diff
@@ -49,4 +49,76 @@ In the example:
...
+
+**Collect poll report**
+Each poll round consists of iterations. In each iteration, the FailureDetector pings all responsive nodes 
+in the layout and also collects their node states to provide cluster state.
+To provide more resilient failre detection mechanism, the failure detector collects few "poll reports" 
+equal to `failureThreshold` number then aggregates final poll report
+
+Once a poll report collected, the `failureDetectionTask` is triggered by Remote Monitoring Service.
+To detect a failure in the cluster the Remote Monitoring Service gets `cluster state` from a poll report.
+The cluster state contains information of the cluster.
```
Comment (PavelZaytsev) [inline docs/failure-detector/failure-detector.md]: `information about`

File: docs/failure-detector/failure-detector.md (hunk 4 above)
Comment (PavelZaytsev) [reply docs/failure-detector/failure-detector.md]: I think we should specifically state here that 'fully connected' in this context means fully connected to the subset of alive nodes

File: docs/failure-detector/failure-detector.md [hunk 9]
```diff
@@ -49,4 +49,76 @@ In the example:
...
+with the smallest number of successful connections and with the highest name (sorted alphabetically).
+
+Even if we have found a node with smallest number of connections it doesnt mean that the node is 
+failed. We also have to check if the node is *fully connected*. 
+Only if the node is not fully connected and has the smallest number of successful connections then
+the node can be added to the unresponsive list.   
+
+**Healing process** - determine if a node is alive and remove it from the unresponsive list.
+The node can heal only itself. The node responsible only for itself, can't heal other nodes.
+This contract simplifies healing algorithm and guaranties that if node became available it mark itself as a responsible
```
Comment (PavelZaytsev) [inline docs/failure-detector/failure-detector.md]: `a node becomes available, it marks`

File: docs/failure-detector/failure-detector.md (hunk 4 above)
Comment (xnull) [reply docs/failure-detector/failure-detector.md]: it is already explained in this docment. See https://github.com/CorfuDB/CorfuDB/blob/master/docs/failure-detector/failure-detector.md
image

File: docs/failure-detector/failure-detector.md (hunk 8 above)
Comment (xnull) [reply docs/failure-detector/failure-detector.md]: done

File: docs/failure-detector/failure-detector.md (hunk 9 above)
Comment (xnull) [reply docs/failure-detector/failure-detector.md]: done

**Pull request merged**